### PR TITLE
Referral onboarding

### LINKED ISSUE #2240: [ADD] Referral System UI/UX
## Overview
Implement a user-friendly system that allows users to register and manage their own THORNames which we will use as referrals, set duration, choose payout assets, and view their referral earnings.

## Background
THORChain supports an affiliate fee system that enables earnings through transaction fees. This feature will implement a user-friendly referral registration and management system that allows users to create and manage their THORNames.

### 1. Create Referral Flow
- User enters their preferred THORName for referral, max. 4 characters
- System validates and checks if the THORName is available
- 10 RUNE THORName registration fee
- User selects the duration for the referral (1 RUNE = 1 year cost)
- User selects preferred payout asset from active THORChain pools
- App displays cost summary and confirmation screen
- App creates the appropriate transaction memo for THORName registration
- User signs the transaction to register their THORName

### 2. Manage Referral Flow
- Display referral name, status, expiration date, and earnings
- Allow editing of preferred payout asset and extending the referral duration

### 3. Save External Referral Flow
- User can input someone else's THORName as a referral code
- System validates that the THORName exists on THORChain
- System verifies that the input THORName is not the user's own referral code
- User can save the referral code to be used in future swap transactions
- User can view and manage saved referral code

### THORName Integration
1. **THORName API Integration**
   - Check THORName availability: `/thorname/lookup/{thorname}`
   - Query existing THORName details: `/thorname/{thorname}`
   - Reverse lookup (address to name): `/thorname/lookup/{address}`
   - Register new THORNames on THORChain via MsgDeposit transaction
   - Query network parameters: `https://thornode.ninerealms.com/thorchain/network`

2. **THORChain Pool Integration**
   - Fetch and display active pools from THORChain
   - Allow selection of preferred payout asset from these pools
   - Validate that selected asset is supported

3. **Referral Management**
   - Store and retrieve user's THORName
   - Track referral expiration date
   - Monitor referral earnings through THORChain API

### Registration Transaction Handling
1. **Memo Generation**
   - Use the following memo format for THORName registration:
     ```
     ~:name:chain:address:?owner:?preferredAsset:?expiry
     ```
     - name: THORName (Just a maximum of 4 characters)
     - chain: The chain of the alias to set
     - address: The alias address (must be a valid address for the chain)
     - owner: THORChain address of owner (optional)
     - preferredAsset: Asset to receive fees in (must be an active pool asset)
     - Example: `~:ODIN:BTC:bc1Address:thorAddress:BTC.BTC`

   - For setting preferred asset, include it in the registration memo:
     ```
     ~:thorname:ETH:ethAddress:thorAddress:ETH.USDC-0XA0B86991C6218B36C1D19D4A2E9EB0CE3606EB48
     ```

   - For renewing a THORName, use:
     ```
     ~:thorname:THOR:<thor-alias-address>
     ```

2. **Fee Calculation**
   - Registration fee: `tns_register_fee_rune` from network endpoint
   - Per block fee: `tns_fee_per_block_rune`
   - Amount calculation for 1 year: `tns_register_fee_rune + (tns_fee_per_block_rune * 5256000)`
   - Renewal amount: `num_blocks_to_extend * tns_fee_per_block_rune`

### Possible Data Models
```swift
struct THORNameReferral {
    let name: String
    let registrationBlockHeight: Int64
    let expirationBlockHeight: Int64
    let preferredAsset: String
    
    // Calculate status by comparing current block height to expiration
    var status: ReferralStatus {
        let currentBlockHeight = NetworkService.shared.getCurrentBlockHeight()
        if expirationBlockHeight < currentBlockHeight {
            return .expired
        } else {
            return .active
        }
    }
    
    // Convert block heights to user-friendly dates for display
    var expirationDate: Date {
        // Average block time is ~6 seconds
        let secondsRemaining = (expirationBlockHeight - NetworkService.shared.getCurrentBlockHeight()) * 6
        return Date().addingTimeInterval(TimeInterval(secondsRemaining))
    }
    
    // Calculate remaining duration in days
    var remainingDays: Int {
        let blocksRemaining = expirationBlockHeight - NetworkService.shared.getCurrentBlockHeight()
        // ~14400 blocks per day (6 seconds per block)
        return Int(blocksRemaining / 14400)
    }
}

// Enum representing THORName status
enum ReferralStatus: String {
    case active = "Active"
    case expired = "Expired"
}

// Earnings can be queried directly from THORChain API
typealias ReferralEarnings = [String: Decimal] // asset -> amount

// Saved referral for use in swaps
struct SavedReferral {
    let name: String           // THORName used as referral
    let isActive: Bool         // Whether this referral is currently active for swaps
}
```

### API Services
1. **THORNameService**
   - `checkAvailability(name: String) -> Bool`
   - `registerName(name: String, duration: Int, preferredAsset: String) -> Transaction`
   - `getReferralStatus(name: String) -> THORNameReferral`
   - `validateExternalReferral(name: String, userAddress: String) -> Bool` // Check THORName exists and isn't owned by user

2. **THORChainPoolService**
   - `getActivePools() -> [Pool]`
   - `getPreferredAssetOptions() -> [Asset]`

3. **ReferralTrackerService**
   - `getReferralEarnings(name: String) -> [ReferralEarnings]`
   - `getTotalEarnings() -> [String: Decimal]` // asset -> amount

### UI Implementation
Please use the whole [figma](https://www.figma.com/design/puB2fsVpPrBx3Sup7gaa3v/Vultisig-App?node-id=38079-21785&t=K24QGxg9e55HXZFz-4) flow to understand the UI/UX.
Entry:

Image

Image

1. **Create Screens**
   - THORName entry and validation screen
   - Duration selection screen
   - Asset selection screen showing active THORChain pools
   - Cost summary and confirmation screen

Image

2. **Management Screens**
   - Referral dashboard showing active referrals
   - Earnings summary with total amounts
   - Edit referral preferences screen

## Testing Plan
1. **Unit Tests**
   - THORName validation and availability checks
   - Registration memo formatting
   - Fee calculations

2. **Integration Tests**
   - THORName registration verification
   - Preferred asset configuration
   - Renewal process verification

3. **UI Tests**
   - Complete user flow for creating a referral
   - Referral management screens
   - Error handling and edge cases

## Implementation Checklist
- [ ] User can enter and validate THORName availability
- [ ] System correctly checks THORName availability via THORChain API
- [ ] User can select referral duration (1 RUNE = 1 year)
- [ ] User can select preferred payout asset from active THORChain pools
- [ ] System correctly displays cost summary with registration fees
- [ ] System generates correctly formatted THORName registration memo
- [ ] Transaction signing process works correctly for THORName registration
- [ ] THORName status and expiration are correctly displayed after registration
- [ ] Referral earnings are displayed correctly
- [ ] User can edit preferred payout asset
- [ ] User can extend THORName duration
- [ ] User can input and save external THORName as referral for swaps
- [ ] System validates saved referral exists on THORChain
- [ ] System prevents users from using their own THORName as a referral
- [ ] User can view and manage saved referral codes
- [ ] Error handling for invalid THORNames or API failures

## Resources
- [THORChain Affiliate Fee Guide](https://dev.thorchain.org/affiliate-guide/affiliate-fee-guide.html)
- [THORName Guide](https://dev.thorchain.org/affiliate-guide/thorname-guide.html)
- [THORChain Name Service Documentation](https://docs.thorchain.org/how-it-works/thorchain-name-service)
- [THORChain Transaction Memo Format](https://dev.thorchain.org/concepts/memos.html)

### PR DESCRIPTION
**Pull request by Rockindash**
## Description

Implementing Referral flow, Part 1. Added the following views:
- UseReferralCodeView
- ReferralOnboardingView

Please Note:
I will be merging the Referral flow into main branch throughout the process to avoid a big PR merge. The link from Settings wont be available until the entire is implemented.

Fixes #2240

### CODE REVIEW

**Review by johnnyluo — APPROVED** [review]
LGTM

File: VultisigApp/VultisigApp/Views/Referral/UseReferralCodeView.swift [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    
+    var title: some View {
+        Text(NSLocalizedString("useReferralCode", comment: ""))
+            .frame(maxWidth: .infinity, alignment: .leading)
+            .font(.body14BrockmannMedium)
+    }
+    
+    var textField: some View {
+        ReferralTextFieldWithCopy(
+            placeholderText: "enterUpto4Characters",
```
[the remark this answers is not in the document]
Comment (Rockindash) [reply VultisigApp/VultisigApp/Views/Referral/UseReferralCodeView.swift:42]: Still a WIP.

File: VultisigApp/VultisigApp/Views/Components/ReferralTextFieldWithCopy.swift [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        .background(Color.blue600)
+        .cornerRadius(12)
+        .overlay(
+            RoundedRectangle(cornerRadius: 12)
+                .stroke(showError ? Color.invalidRed : Color.blue200, lineWidth: 1)
+        )
+        .autocorrectionDisabled()
+        .borderlessTextFieldStyle()
+        .padding(1)
+    }
```
[the remark this answers is not in the document]
Comment (Rockindash) [reply VultisigApp/VultisigApp/Views/Components/ReferralTextFieldWithCopy.swift:47]: Still a WIP.

File: VultisigApp/VultisigApp/Views/Components/ReferralTextFieldWithCopy.swift [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        .cornerRadius(12)
+        .overlay(
+            RoundedRectangle(cornerRadius: 12)
+                .stroke(showError ? Color.invalidRed : Color.blue200, lineWidth: 1)
+        )
+        .textInputAutocapitalization(.never)
+        .autocorrectionDisabled()
+        .borderlessTextFieldStyle()
+        .padding(1)
+    }
```
[the remark this answers is not in the document]
Comment (Rockindash) [reply VultisigApp/VultisigApp/Views/Components/ReferralTextFieldWithCopy.swift]: Still a WIP.
